README.md
# @davidgovea/react-native-wheel-datepicker

[![NPM](https://img.shields.io/npm/v/@davidgovea/react-native-wheel-datepicker.svg)](https://www.npmjs.com/package/@davidgovea/react-native-wheel-datepicker)
[![Build Status](https://img.shields.io/circleci/project/github/davidgovea/react-native-wheel-datepicker.svg)](https://circleci.com/gh/davidgovea/react-native-wheel-datepicker)
[![](https://img.shields.io/codecov/c/github/davidgovea/react-native-wheel-datepicker.svg)](https://codecov.io/gh/davidgovea/react-native-wheel-datepicker)

A iOS style picker and datepicker components for ReactNative.

- Original repository by @lesliesam: [react-native-wheel-picker](https://github.com/lesliesam/react-native-wheel-picker).
- Fork by @wusuopu who added the real cross platform behavior and datepicker [react-native-wheel-picker](https://github.com/wusuopu/react-native-wheel-picker).
- Fork by @pinguinjkeke, who provided maintenance and fixes [react-native-wheel-datepicker](https://github.com/pinguinjkeke/react-native-wheel-datepicker)

This is the fourth fork of repository, since it seems that @pinguinjkeke is no longer supporting [react-native-wheel-datepicker](https://github.com/pinguinjkeke/react-native-wheel-datepicker).

### Goals of this fork:
* Maintenance / updates for RN & platform changes
* Implement some missing features on Android (12-hour AM/PM mode, minute intervals)
* Typescript definitions
* Add tests
* Semver adherance

## Introduction
Cross platform Picker component for React-native.

Since picker is originally supported by iOS while Android only supports a ugly Spinner component. If you want to have the same user behaviour, you can use this.

The android component is based on [wheel-picker](https://github.com/AigeStudio/WheelPicker) which runs super fast and smoothly. It also supports curved effect which make it exactly the same looking and feel as the ios picker.

![](screenshots/ios1.png)
![](screenshots/ios2.png)

![](screenshots/android.png)

## Installation

```
npm i @davidgovea/react-native-wheel-datepicker --save
```

Configration (for RN below 0.60):

```
react-native link @davidgovea/react-native-wheel-datepicker
```

## &lt;Picker&gt; Component

```jsx
import { Picker } from '@davidgovea/react-native-wheel-datepicker';

<Picker
  style={{ flex: 1 }}
  selectedValue={3}
  pickerData={[1, 2, 3, 4, 5, 6]}
  onValueChange={value => { console.log(value); }}
/>
```

### Picker props

| Property      | Type                 | Default                    | Description                                         |
|---------------|----------------------|----------------------------|-----------------------------------------------------|
| pickerData    | any[]                | -                          | **Required:** an array of data to display in picker |
| onValueChange | (value: any) => void | -                          | **Required:** selected value changed callback       |
| selectedValue | any?                 | First item of `pickerData` | Set/control the selected value                      |
| textColor     | string               | '#333'                     | (Android only) Color of spinner text                |
| textSize      | number               | 26                         | (Android only) Spinner text size                    |
| itemSpace     | number               | 20                         | (Android only) Spacing between spinner items        |
| itemStyle     | style                | null                       | (iOS only) passed into `PickerIOS` for item styling |
| style         | style                | null                       | Style applied to the picker element                 |

---

## &lt;DatePicker&gt; Component

```jsx
import { DatePicker } from '@davidgovea/react-native-wheel-datepicker';

<DatePicker
  mode="datetime"
  use12Hours
  minuteInterval={15}
  onDateChange={(date) => { console.log(date); }}
/>

```

### DatePicker props
| Property       | Type                                          | Default                           | Description                                                                                                                              |
|----------------|-----------------------------------------------|-----------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
| onDateChange   | (value: Date) => void                         | -                                 | **Required:** selected date changed callback                                                                                             |
| mode           | enum('date', 'time', 'datetime')              | 'date'                            | Show date, time, or both in picker                                                                                                       |
| date           | Date                                          | `new Date()` (now)                | Provide/control the selected date value                                                                                                  |
| maximumDate    | Date                                          | 10 years in future                | Restricts the range of possible date/time values.                                                                                        |
| minimumDate    | Date                                          | 10 years in past                  | Restricts the range of possible date/time values.                                                                                        |
| minuteInterval | enum(1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30)    | 1                                 | The interval at which minutes can be selected.                                                                                           |
| use12Hours     | boolean                                       | false                             | (Android only) Use 12-hour AM/PM for hour selection. On iOS, `DatePickerIOS` automatically uses i18n information to enable 12-hour mode. |
| order          | custom string                                 | 'D-M-Y'                           | (Android only) Ordering of date fields. Use "D" for day, "M" for month, "Y" for year, separated by dashes ("-")                          |
| labelUnit      | { year: string, month: string, date: string } | { year: '', month: '', date: '' } | (Android only) Suffix labels for year, month, date values                                                                                |
| textColor      | string                                        | '#333'                            | (Android only) Color of spinner text                                                                                                     |
| textSize       | number                                        | 26                                | (Android only) Spinner text size                                                                                                         |
| itemSpace      | number                                        | 20                                | (Android only) Spacing between spinner items                                                                                             |
| style          | style                                         | null                              | Style applied to the datepicker element                                                                                                  |

## [DEPRECATED] Integration with CustomDatePickerIOS

> CustomDatePickerIOS can apparently cause Appstore approval issues. This library will be removing iOS functionality in a future release. Users are advised to use the builtin Pickers on iOS.

By default, package provides default DatePickerIOS on the iOS side to simplify usage on both platforms.

You can install [react-native-custom-datepicker-ios](https://github.com/pinguinjkeke/react-native-custom-datepicker-ios) package
if you need textColor functionality on IOS platform.

Just install:
```
npm i react-native-custom-datepicker-ios
// or
yarn add react-native-custom-datepicker-ios
```
link dependencies:
```
react-native link react-native-custom-datepicker-ios
```
And register CustomDatePickerIOS inside react-native-wheel-datepicker package.
```js
import { registerCustomDatePickerIOS } from '@davidgovea/react-native-wheel-datepicker';
import CustomDatePickerIOS from 'react-native-custom-datepicker-ios';

registerCustomDatePickerIOS(CustomDatePickerIOS);
```
Then you can use textColored components for both platforms inside render function!
```jsx
import { DatePicker } from '@davidgovea/react-native-wheel-datepicker';

// ...
render() {
  return (
    <DatePicker
      mode="date"
      textColor="green"
    />
  )
}
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v1.0.1...v1.0.2) (2020-05-29)


### Bug Fixes

* Address deprecated gradle config ([c8e7b56](https://github.com/davidgovea/react-native-wheel-datepicker/commit/c8e7b567d72778d3c164abb8c9d39c89c15121c7)), closes [#18](https://github.com/davidgovea/react-native-wheel-datepicker/issues/18)

### [1.0.1](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v1.0.0...v1.0.1) (2019-12-29)


### Bug Fixes

* omit uneeded files from npm archive ([425a0e3](https://github.com/davidgovea/react-native-wheel-datepicker/commit/425a0e39c166446ef4e292713cb3287f9f137a67))

## [1.0.0](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.232...v1.0.0) (2019-12-29)


### Features

* Add 12-hour timepicker example to Demo app ([6f40500](https://github.com/davidgovea/react-native-wheel-datepicker/commit/6f405009003706e5bbcacbb10233a56ec43eeb03))
* Document props & usage in README ([6fb2fac](https://github.com/davidgovea/react-native-wheel-datepicker/commit/6fb2facaf284dde32f7dc7d715d5c4fcf0968293))
* remove deprecated custom ios picker extension from Demo app ([05a7f1d](https://github.com/davidgovea/react-native-wheel-datepicker/commit/05a7f1d5218f3a5ec7554ada1695c757bf006800))
* Switch Demo app from yarn to npm ([cda526f](https://github.com/davidgovea/react-native-wheel-datepicker/commit/cda526f872fa4cf0fd527ebee013fea3c75d763b))


### Bug Fixes

* Remove duplicate package import in android Demo app ([74e8a77](https://github.com/davidgovea/react-native-wheel-datepicker/commit/74e8a77304241d638ec5e476cdf477f0289e6bc1))
* Remove unneeded moment peerDependency. Moment is a full-on dependency already ([f883338](https://github.com/davidgovea/react-native-wheel-datepicker/commit/f8833384c71b4b0ffbdd5efad384aa6c57501533))
* Scrollview height bug - android was glitching due to flex:1 limiting the height ([834f8ad](https://github.com/davidgovea/react-native-wheel-datepicker/commit/834f8ad6b10b86203092438a2a21fb9420bcc48d))
* Update package names in Demo app ([d6129dd](https://github.com/davidgovea/react-native-wheel-datepicker/commit/d6129ddae2f86167edbbf02473620088a8bf265e))

### [0.0.232](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.231...v0.0.232) (2019-12-28)


### Bug Fixes

* use "ViewProps.style" for the `style` type definitions ([#12](https://github.com/davidgovea/react-native-wheel-datepicker/issues/12)) ([73ae46a](https://github.com/davidgovea/react-native-wheel-datepicker/commit/73ae46a431babf66dc018b52ee92509499928154))

### [0.0.231](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.230...v0.0.231) (2019-12-08)


### Bug Fixes

* React deprecations and readme/demo app instructions ([f6dc3b2](https://github.com/davidgovea/react-native-wheel-datepicker/commit/f6dc3b2b6d51a4f4aa8afab0c200cfe73c3e991a))

### [0.0.230](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.229...v0.0.230) (2019-11-03)


### Features

* add typescript definition ([#9](https://github.com/davidgovea/react-native-wheel-datepicker/issues/9)) ([53ff827](https://github.com/davidgovea/react-native-wheel-datepicker/commit/53ff8279d1ff5a88b1d9c751a35208be717d21c8))

### [0.0.229](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.228...v0.0.229) (2019-11-03)


### Bug Fixes

* use proper itemStyle prop definition ([d1232ed](https://github.com/davidgovea/react-native-wheel-datepicker/commit/d1232ed8088510bc0a81873e5a9a291218d574d3)), closes [pinguinjkeke/react-native-wheel-datepicker#31](https://github.com/pinguinjkeke/react-native-wheel-datepicker/issues/31)

### [0.0.228](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.227...v0.0.228) (2019-11-03)


### Features

* Added 12-hour AM/PM mode for Android ([6f13ad5](https://github.com/davidgovea/react-native-wheel-datepicker/commit/6f13ad57cdd4be9674900097c595878b0af31aca))

### [0.0.227](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.226...v0.0.227) (2019-11-03)


### Features

* add ci builds, coverage, and initial test ([067be9c](https://github.com/davidgovea/react-native-wheel-datepicker/commit/067be9c6289e5be03462540daaa46ce6b7f34a99))
* basic coverage via snapshot & shallow render tests ([8d54fed](https://github.com/davidgovea/react-native-wheel-datepicker/commit/8d54fedcaed87040ff96839b85bc313c0a5c3220))


### Bug Fixes

* oneOf proptypes syntax error ([9f88863](https://github.com/davidgovea/react-native-wheel-datepicker/commit/9f88863da1e0a17c757a87ace34da64c211a192a))

### [0.0.226](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.225...v0.0.226) (2019-11-02)


### Features

* implement `minuteInterval` on android ([1ac13dc](https://github.com/davidgovea/react-native-wheel-datepicker/commit/1ac13dc540c9985aa7a4cc0e762cafca0082e72f))

### [0.0.225](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.224...v0.0.225) (2019-11-02)


### Bug Fixes

* rnpm deprecation ([7636648](https://github.com/davidgovea/react-native-wheel-datepicker/commit/76366487dfc69ddc52ff639fb60731e1146cd4f1))

## [0.0.224](https://github.com/davidgovea/react-native-wheel-datepicker/compare/v0.0.223...v0.0.224) (2019-11-02)


### Bug Fixes

* more current default for `compileSdkVersion` on android ([40da618](https://github.com/davidgovea/react-native-wheel-datepicker/commit/40da618))


## 0.0.223 (2019-11-02)


* Forked repo from [react-native-wheel-datepicker](https://github.com/pinguinjkeke/react-native-wheel-datepicker)

src/date-picker.android.js
import React, { PureComponent } from 'react';
import { ColorPropType, StyleSheet, View, ViewPropTypes as RNViewPropTypes } from 'react-native';
import PropTypes from 'prop-types';
import moment from 'moment';
import Picker from './picker';

const ViewPropTypes = RNViewPropTypes || View.propTypes;

const styles = StyleSheet.create({
  picker: {
    flex: 1,
  },
  row: {
    flexDirection: 'row',
  },
});

const stylesFromProps = props => ({
  itemSpace: props.itemSpace,
  textColor: props.textColor,
  textSize: props.textSize,
  style: props.style,
});

export default class DatePicker extends PureComponent {
  static propTypes = {
    labelUnit: PropTypes.shape({
      year: PropTypes.string,
      month: PropTypes.string,
      date: PropTypes.string,
    }),
    order: PropTypes.string,
    date: PropTypes.instanceOf(Date).isRequired,
    maximumDate: PropTypes.instanceOf(Date),
    minimumDate: PropTypes.instanceOf(Date),
    mode: PropTypes.oneOf(['date', 'time', 'datetime']),
    onDateChange: PropTypes.func.isRequired,
    use12Hours: PropTypes.bool,
    minuteInterval: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30]),
    style: ViewPropTypes.style,
    textColor: ColorPropType,
    textSize: PropTypes.number,
    itemSpace: PropTypes.number,
  };

  static defaultProps = {
    labelUnit: { year: '', month: '', date: '' },
    order: 'D-M-Y',
    mode: 'date',
    maximumDate: moment().add(10, 'years').toDate(),
    minimumDate: moment().add(-10, 'years').toDate(),
    use12Hours: true,
    date: new Date(),
    style: null,
    textColor: '#333',
    textSize: 26,
    itemSpace: 20,
  };

  constructor(props) {
    super(props);

    const { date, minimumDate, maximumDate, labelUnit } = props;

    this.state = { date, monthRange: [], yearRange: [] };

    this.newValue = {};

    this.parseDate(date);

    const mdate = moment(date);

    const dayNum = mdate.daysInMonth();
    this.state.dayRange = this.genDateRange(dayNum);

    const minYear = minimumDate.getFullYear();
    const maxYear = maximumDate.getFullYear();

    for (let i = 1; i <= 12; i += 1) {
      this.state.monthRange.push({ value: i, label: `${i}${labelUnit.month}` });
    }

    this.state.yearRange.push({ value: minYear, label: `${minYear}${labelUnit.year}` });

    for (let i = minYear + 1; i <= maxYear; i += 1) {
      this.state.yearRange.push({ value: i, label: `${i}${labelUnit.year}` });
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.state.date !== nextProps.date) {
      this.parseDate(nextProps.date);

      this.setState({ date: nextProps.date });
    }
  }

  parseDate = (date) => {
    const mdate = moment(date);

    ['year', 'month', 'date', 'hour', 'minute'].forEach((s) => { this.newValue[s] = mdate.get(s); });
  }

  get meridiem() {
    return this.newValue.hour >= 12 ? 'PM' : 'AM';
  }

  onYearChange = (year) => {
    const oldYear = this.newValue.year;

    this.newValue.year = year;
    this.checkDate(oldYear, this.newValue.month);
    this.props.onDateChange(this.getValue());
  };

  onMonthChange = (month) => {
    const oldMonth = this.newValue.month;

    this.newValue.month = month - 1;
    this.checkDate(this.newValue.year, oldMonth);
    this.props.onDateChange(this.getValue());
  };

  onDateChange = (date) => {
    this.newValue.date = date;
    this.checkDate(this.newValue.year, this.newValue.month);
    this.props.onDateChange(this.getValue());
  };

  onHourChange = (hour) => {
    if (this.props.use12Hours && hour < 12 && this.meridiem === 'PM') {
      // +12hrs for PM times (except 12)
      this.newValue.hour = hour + 12;
    } else if (this.props.use12Hours && hour === 12 && this.meridiem === 'AM') {
      // 12 AM -> 0:00
      this.newValue.hour = 0;
    } else {
      this.newValue.hour = hour;
    }

    this.props.onDateChange(this.getValue());
  };

  onMinuteChange = (minute) => {
    this.newValue.minute = minute;
    this.props.onDateChange(this.getValue());
  };

  onPeriodChange = (period) => {
    if (this.meridiem !== period) {
      let newHour = (this.newValue.hour + 12) % 24;
      if (period === 'AM') {
        newHour %= 12;
      }
      this.newValue.hour = newHour;
      this.props.onDateChange(this.getValue());
    }
  };

  genDateRange(dayNum) {
    const days = [];

    for (let i = 1; i <= dayNum; i += 1) {
      days.push({ value: i, label: `${i}${this.props.labelUnit.date}` });
    }

    return days;
  }

  render() {
    return (
      <View style={styles.row}>
        {['date', 'datetime'].includes(this.props.mode) && this.datePicker}
        {['time', 'datetime'].includes(this.props.mode) && this.timePicker}
      </View>
    );
  }

  get datePicker() {
    const propsStyles = stylesFromProps(this.props);

    const { order } = this.props;

    if (!order.includes('D') && !order.includes('M') && !order.includes('Y')) {
      throw new Error(`WheelDatePicker: you are using order prop wrong, default value is 'D-M-Y'`);
    }

    return this.props.order.split('-').map((key) => {
      switch (key) {
        case 'D': return (
          <View key='date' style={styles.picker}>
            <Picker
              {...propsStyles}
              style={this.props.style}
              ref={(date) => { this.dateComponent = date; }}
              selectedValue={this.state.date.getDate()}
              pickerData={this.state.dayRange}
              onValueChange={this.onDateChange}
            />
          </View>
        );
        case 'M': return (
          <View key='month' style={styles.picker}>
            <Picker
              {...propsStyles}
              style={this.props.style}
              ref={(month) => { this.monthComponent = month; }}
              selectedValue={this.state.date.getMonth() + 1}
              pickerData={this.state.monthRange}
              onValueChange={this.onMonthChange}
            />
          </View>
        );
        case 'Y': return (
          <View key='year' style={styles.picker}>
            <Picker
              {...propsStyles}
              style={this.props.style}
              ref={(year) => { this.yearComponent = year; }}
              selectedValue={this.state.date.getFullYear()}
              pickerData={this.state.yearRange}
              onValueChange={this.onYearChange}
            />
          </View>
        );
        default: return null;
      }
    });
  }

  get timePicker() {
    const propsStyles = stylesFromProps(this.props);
    const { minuteInterval, use12Hours } = this.props;

    const [hours, minutes] = [[], []];

    const maxHours = use12Hours ? 12 : 23;
    const minHours = use12Hours ? 1 : 0;
    let initialHours = this.state.date.getHours();
    if (use12Hours && initialHours > 12) initialHours -= 12;

    for (let i = minHours; i <= maxHours; i += 1) {
      hours.push(i);
    }

    for (let i = 0; i <= 59; i += 1) {
      if (!minuteInterval || i % minuteInterval === 0) {
        minutes.push({
          value: i,
          label: i > 9 ? `${i}` : `0${i}`,
        });
      }
    }

    return [
      <View key='hour' style={styles.picker}>
        <Picker
          ref={(hour) => { this.hourComponent = hour; }}
          {...propsStyles}
          selectedValue={initialHours}
          pickerData={hours}
          onValueChange={this.onHourChange}
        />
      </View>,
      <View key='minute' style={styles.picker}>
        <Picker
          ref={(minute) => { this.minuteComponent = minute; }}
          {...propsStyles}
          selectedValue={this.state.date.getMinutes()}
          pickerData={minutes}
          onValueChange={this.onMinuteChange}
        />
      </View>,
      use12Hours && <View key='period' style={styles.picker}>
        <Picker
          ref={(period) => { this.periodComponent = period; }}
          {...propsStyles}
          selectedValue={moment(this.state.date).format('A')}
          pickerData={['AM', 'PM']}
          onValueChange={this.onPeriodChange}
        />
      </View>,
    ];
  }

  checkDate(oldYear, oldMonth) {
    const currentMonth = this.newValue.month;
    const currentYear = this.newValue.year;
    const currentDay = this.newValue.date;

    let { dayRange } = this.state;
    let dayNum = dayRange.length;

    if (oldMonth !== currentMonth || oldYear !== currentYear) {
      dayNum = moment(`${currentYear}-${currentMonth + 1}`, 'YYYY-MM').daysInMonth();
    }

    if (dayNum !== dayRange.length) {
      dayRange = this.genDateRange(dayNum);

      if (currentDay > dayNum) {
        this.newValue.date = dayNum;
        this.dateComponent.setState({ selectedValue: dayNum });
      }

      this.setState({ dayRange });
    }

    const unit = this.props.mode === 'date' ? 'day' : undefined;
    const current = Object.assign({}, this.newValue, { date: this.newValue.date });
    let currentTime = moment(current);
    const min = moment(this.props.minimumDate);
    const max = moment(this.props.maximumDate);
    let isCurrentTimeChanged = false;

    if (currentTime.isBefore(min, unit)) {
      [currentTime, isCurrentTimeChanged] = [min, true];
    } else if (currentTime.isAfter(max, unit)) {
      [currentTime, isCurrentTimeChanged] = [max, true];
    }

    if (isCurrentTimeChanged) {
      if (this.monthComponent) {
        this.monthComponent.setState({ selectedValue: currentTime.get('month') + 1 });
      }

      ['year', 'date', 'hour', 'minute'].forEach((segment) => {
        const ref = this[`${segment}Component`];

        return ref && ref.setState({ selectedValue: currentTime.get(segment) });
      });
    }
  }

  getValue() {
    const { year, month, date, hour, minute } = this.newValue;
    const nextDate = new Date(year, month, date, hour, minute);

    if (nextDate < this.props.minimumDate) {
      return this.props.minimumDate;
    }

    return nextDate > this.props.maximumDate ? this.props.maximumDate : nextDate;
  }
}
